### PR TITLE
Bug 5036

### PR DESCRIPTION
**Pull request by mattam82**
Fix for bug #5036 (on top of previous bug, which should disappear when merged).

### CODE REVIEW

File: library/declare.mli [hunk 1]
```diff
@@ -87,7 +87,9 @@ val exists_name : Id.t -> bool
 
 
 
-(** Global universe names and constraints *)
+(** Global universe contexts, names and constraints *)
 
+val declare_universe_context : polymorphic -> Univ.universe_context_set -> unit
```
Comment (maximedenes) [inline library/declare.mli]: Why is this not called do_universe_context for uniformity? What does polymorphic mean exactly? Also, it should be a label. We already have too many functions taking unlabeled booleans as arguments, please don't add new ones.

File: ltac/extratactics.ml4 [hunk 2]
```diff
@@ -263,7 +263,8 @@ let add_rewrite_hint bases ort t lcsr =
     let ctx =
       let ctx = UState.context_set ctx in
         if poly then ctx
-	else (Global.push_context_set false ctx; Univ.ContextSet.empty)
+	else (Declare.declare_universe_context false ctx;
```
Comment (maximedenes) [inline ltac/extratactics.ml4]: Can you comment a bit what the code does here?

File: library/declare.ml [hunk 3]
```diff
@@ -475,8 +491,8 @@ let do_universe poly l =
 type constraint_decl = polymorphic * Univ.constraints
 
 let cache_constraints (na, (p, c)) =
-  Global.add_constraints c;
-  if p then Lib.add_section_context (Univ.ContextSet.add_constraints c Univ.ContextSet.empty)
+  let ctx = Univ.ContextSet.add_constraints c Univ.ContextSet.empty in
```
Comment (maximedenes) [inline library/declare.ml]: Can you comment why an empty ContextSet is what we want here?

Comment (mattam82) [inline library/declare.ml]: We use a general declare_universe_context, that do_constraints and do_universes rely on. One pushes just constraints, the other just universes.

File: library/declare.mli (hunk 1 above)
Comment (mattam82) [inline library/declare.mli]: I'll leave a general labelization of ~polymorphic for later (it should rather be done everywhere at once)

**Pull request merged**